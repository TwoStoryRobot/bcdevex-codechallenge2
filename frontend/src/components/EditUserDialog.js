import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Dialog from '@material-ui/core/Dialog'
import DialogTitle from '@material-ui/core/DialogTitle'
import TextField from '@material-ui/core/TextField'
import DialogContent from '@material-ui/core/DialogContent'
import DialogActions, from '@material-ui/core/DialogActions'
import Button from '@material-ui/core/Button'
import styled from 'styled-components'

const StyledDialogContent = styled(DialogContent)`
  display: grid;
  grid-template-columns: auto auto;
  grid-column-gap: 20px;
`

const AvatarWrapper = styled.div`
  grid-column: 1;
`

const StyledForm = styled.form`
  grid-column: 2;
`

export default class EditUserDialog extends Component {
  constructor(props) {
    super(props)

    const { firstName, lastName, avatarUrl, email } = props
    this.state = {
      firstName,
      lastName,
      email,
      avatarUrl
    }
  }

  handleTextFieldChange = name => event => {
    this.setState({
      [name]: event.target.value,
    })
  }

  handleCheckboxChange = name => event => {
    this.setState({
      [name]: event.target.checked
    })
  }

  handleSave = () => {
    const { firstName, lastName, email, avatarUrl } = this.state

    this.props.onSave({ firstName, lastName, email, avatarUrl })
  }
  
  handleKeyDown = event => {
    if (event.keyCode === 13) {
      this.handleSave()
    }
  }

  render() {
    const { open, onClose } = this.props
    const { firstName, lastName, avatarUrl, email } = this.state
  
    return (
      <Dialog open={open} onClose={onClose} data-testid="edit-user-dialog">
        <DialogTitle>Edit User</DialogTitle>
        <StyledDialogContent>
          <AvatarWrapper>
            <Avatar
              src={avatarUrl}
              name={`${firstName} ${lastName}`}
              round
              size={40}
            />
          </AvatarWrapper>
          <StyledForm onSubmit={this.handleSave}>
            <TextField
              id="first"
              label="First"
              value={firstName}
              onChange={this.handleTextFieldChange('firstName')}
              onKeyDown={this.handleKeyDown}
              fullWidth
              margin="dense"
              autoFocus
            />
            <TextField
              id="last"
              label="Last"
              value={lastName}
              onChange={this.handleTextFieldChange('lastName')}
              onKeyDown={this.handleKeyDown}
              fullWidth
              margin="dense"
            />
            <TextField
              id="email"
              label="Email"
              value={email}
              onChange={this.handleTextFieldChange('email')}
              onKeyDown={this.handleKeyDown}
              type="email"
              fullWidth
              margin="dense"
            />
            <TextField
              id="avatar"
              label="Avatar URL"
              value={avatarUrl}
              onChange={this.handleTextFieldChange('avatarUrl')}
              onKeyDown={this.handleKeyDown}
              fullWidth
              margin="dense"
            />
          </StyledForm>
        </StyledDialogContent>
        <DialogActions>
          <Button onClick={onClose} data-testid="close">Cancel</Button>
          <Button onClick={this.handleSave} color="primary" data-testid="save">Save</Button>
        </DialogActions>
      </Dialog>
    )
  }
}

EditUserDialog.propTypes = {
  onClose: PropTypes.func,
  onSave: PropTypes.func,
  open: PropTypes.bool,
  firstName: PropTypes.string,
  lastName: PropTypes.string,
  email: PropTypes.string,
  avatarUrl: PropTypes.string,
}

EditUserDialog.defaultProps = {
  firstName: '',
  lastName: '',
  email: '',
  avatarUrl: '',
  open: false,
  onClose: () => {},
  onSave: () => {},
}