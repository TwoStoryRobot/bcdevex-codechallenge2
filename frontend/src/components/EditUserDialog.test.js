import React from 'react'
import {
  cleanup,
  render,
  bindElementToQueries,
  fireEvent
} from 'react-testing-library'

import EditUserDialog from './EditUserDialog'

import 'jest-dom/extend-expect'

afterEach(cleanup)

describe('EditUserDialog', () => {
  it('should only be shown when open=true', () => {
    const { queryByTestId, getByTestId } = bindElementToQueries(document.body)

    const { rerender } = render(<EditUserDialog open={false} />)
    expect(queryByTestId('edit-user-dialog')).toBeNull()

    rerender(<EditUserDialog open={true} />)
    expect(getByTestId('edit-user-dialog')).toBeInTheDocument()
  })

  it('should save any data the user entered when the save button is pressed', () => {
    const save = jest.fn()
    const { getByLabelText, getByTestId } = render(
      <EditUserDialog open={true} onSave={save} />
    )

    const firstName = 'John'
    const lastName = 'Doe'

    const firstInput = getByLabelText('First')
    const lastInput = getByLabelText('Last')

    fireEvent.change(firstInput, { target: { value: firstName } })
    fireEvent.change(lastInput, { target: { value: lastName } })

    expect(firstInput.value).toEqual(firstName)
    expect(lastInput.value).toEqual(lastName)

    const button = getByTestId('save')
    fireEvent.click(button)

    expect(save).toHaveBeenCalledWith({
      firstName,
      lastName,
      emailAddress: '',
      imageURL: ''
    })
  })

  it('Closes without saving when the close button is pressed', () => {
    const save = jest.fn()
    const close = jest.fn()
    const { getByTestId } = render(
      <EditUserDialog open={true} onSave={save} onClose={close} />
    )

    const button = getByTestId('close')
    fireEvent.click(button)

    expect(save).not.toHaveBeenCalled()
    expect(close).toHaveBeenCalled()
  })

  it('Opens with values if the user is editing existing details', () => {
    const firstName = 'John'
    const lastName = 'Doe'
    const imageURL = 'https://api.adorable.io/avatars/285/abott@adorable.png'
    const emailAddress = 'john.doe@example.com'

    const { getByLabelText } = render(
      <EditUserDialog
        open={true}
        firstName={firstName}
        lastName={lastName}
        imageURL={imageURL}
        emailAddress={emailAddress}
      />
    )

    const firstInput = getByLabelText('First')
    const lastInput = getByLabelText('Last')
    const avatarInput = getByLabelText('Avatar URL')
    const emailInput = getByLabelText('Email')

    expect(firstInput.value).toEqual(firstName)
    expect(lastInput.value).toEqual(lastName)
    expect(avatarInput.value).toEqual(imageURL)
    expect(emailInput.value).toEqual(emailAddress)
  })

  it('should match the snapshot', () => {
    const { baseElement } = render(<EditUserDialog open={true} />)
    expect(baseElement).toMatchSnapshot()
  })
})
