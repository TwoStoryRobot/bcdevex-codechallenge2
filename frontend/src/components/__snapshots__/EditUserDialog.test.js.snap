// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUserDialog should match the snapshot 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiModal-root-0561 MuiDialog-root-0549 MuiDialog-scrollPaper-0550"
    role="dialog"
    data-testid="edit-user-dialog"
  >
    <div
      class="MuiBackdrop-root-0563"
      aria-hidden="true"
      style="opacity: 1; -webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      class="MuiPaper-root-0565 MuiPaper-elevation24-0591 MuiPaper-rounded-0566 MuiDialog-paper-0552 MuiDialog-paperScrollPaper-0553 MuiDialog-paperWidthSm-0556"
      style="opacity: 1; -webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      role="document"
      tabindex="-1"
    >
      <div
        class="MuiDialogTitle-root-0592"
      >
        <h2
          class="MuiTypography-root-0593 MuiTypography-title-0599"
        >
          Edit User
        </h2>
      </div>
      <div
        class="MuiDialogContent-root-0618 sc-bdVaJa dhQzIv"
      >
        <div
          class="sc-bwzfXH jDNgdt"
        >
          <div
            class=" sb-avatar"
            style="display: inline-block; vertical-align: middle; width: 40px; height: 40px; border-radius: 100%; font-family: Helvetica, Arial, sans-serif;"
          >
            <div
              class=" sb-avatar__text"
              style="width: 40px; height: 40px; font-size: 13.3333px; line-height: 40px; text-align: center; text-transform: uppercase; color: rgb(255, 255, 255); border-radius: 100%;"
            />
          </div>
        </div>
        <form
          class="sc-htpNat ibEjym"
        >
          <div
            class="MuiFormControl-root-0619 MuiFormControl-fullWidth-0622"
          >
            <div
              class="MuiFormControl-root-0619 MuiFormControl-marginDense-0621 MuiFormControl-fullWidth-0622"
            >
              <label
                class="MuiFormLabel-root-0628 MuiFormLabel-focused-0629 MuiFormLabel-required-0633 MuiInputLabel-root-0623 MuiInputLabel-formControl-0624 MuiInputLabel-animated-0627 MuiInputLabel-shrink-0626 MuiInputLabel-marginDense-0625"
                data-shrink="true"
                for="first"
              >
                First
                <span
                  class="MuiFormLabel-asterisk-0634"
                >
                  *
                </span>
              </label>
              <div
                class="MuiInput-root-0635 MuiInput-fullWidth-0642 MuiInput-focused-0637 MuiInput-formControl-0636 MuiInput-underline-0639"
              >
                <input
                  aria-invalid="false"
                  class="MuiInput-input-0643 MuiInput-inputMarginDense-0644"
                  id="first"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="MuiFormControl-root-0619 MuiFormControl-marginDense-0621 MuiFormControl-fullWidth-0622"
          >
            <label
              class="MuiFormLabel-root-0628 MuiInputLabel-root-0623 MuiInputLabel-formControl-0624 MuiInputLabel-animated-0627 MuiInputLabel-marginDense-0625"
              data-shrink="false"
              for="last"
            >
              Last
            </label>
            <div
              class="MuiInput-root-0635 MuiInput-fullWidth-0642 MuiInput-formControl-0636 MuiInput-underline-0639"
            >
              <input
                aria-invalid="false"
                class="MuiInput-input-0643 MuiInput-inputMarginDense-0644"
                id="last"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root-0619 MuiFormControl-fullWidth-0622"
          >
            <div
              class="MuiFormControl-root-0619 MuiFormControl-marginDense-0621 MuiFormControl-fullWidth-0622"
            >
              <label
                class="MuiFormLabel-root-0628 MuiInputLabel-root-0623 MuiInputLabel-formControl-0624 MuiInputLabel-animated-0627 MuiInputLabel-marginDense-0625"
                data-shrink="false"
                for="emailAddress"
              >
                Email
              </label>
              <div
                class="MuiInput-root-0635 MuiInput-fullWidth-0642 MuiInput-formControl-0636 MuiInput-underline-0639"
              >
                <input
                  aria-invalid="false"
                  class="MuiInput-input-0643 MuiInput-inputType-0646 MuiInput-inputMarginDense-0644"
                  id="emailAddress"
                  type="emailAddress"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="MuiFormControl-root-0619 MuiFormControl-fullWidth-0622"
          >
            <div
              class="MuiFormControl-root-0619 MuiFormControl-marginDense-0621 MuiFormControl-fullWidth-0622"
            >
              <label
                class="MuiFormLabel-root-0628 MuiInputLabel-root-0623 MuiInputLabel-formControl-0624 MuiInputLabel-animated-0627 MuiInputLabel-marginDense-0625"
                data-shrink="false"
                for="avatar"
              >
                Avatar URL
              </label>
              <div
                class="MuiInput-root-0635 MuiInput-fullWidth-0642 MuiInput-formControl-0636 MuiInput-underline-0639"
              >
                <input
                  aria-invalid="false"
                  class="MuiInput-input-0643 MuiInput-inputMarginDense-0644"
                  id="avatar"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </form>
      </div>
      <div
        class="MuiDialogActions-root-0648"
      >
        <button
          tabindex="0"
          class="MuiButtonBase-root-0676 MuiButton-root-0650 MuiButton-text-0652 MuiButton-flat-0655 MuiDialogActions-action-0649"
          type="button"
          data-testid="close"
        >
          <span
            class="MuiButton-label-0651"
          >
            Cancel
          </span>
          <span
            class="MuiTouchRipple-root-0679"
          />
        </button>
        <button
          tabindex="0"
          class="MuiButtonBase-root-0676 MuiButton-root-0650 MuiButton-text-0652 MuiButton-textPrimary-0653 MuiButton-flat-0655 MuiButton-flatPrimary-0656 MuiDialogActions-action-0649"
          type="button"
          data-testid="save"
        >
          <span
            class="MuiButton-label-0651"
          >
            Save
          </span>
          <span
            class="MuiTouchRipple-root-0679"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
